src/storage/page_buffer.c

```diff
@@ -8807,7 +8807,11 @@ pgbuf_put_bcb_into_invalid_list (THREAD_ENTRY * thread_p, PGBUF_BCB * bufptr)
   /* the caller is holding bufptr->mutex */
   VPID_SET_NULL (&bufptr->vpid);
   bufptr->latch_mode = PGBUF_LATCH_INVALID;
-  assert ((bufptr->flags & PGBUF_BCB_FLAGS_MASK) == 0);
+
+  assert ((bufptr->flags & PGBUF_BCB_FLAGS_MASK) == 0
+	  || (is_passive_transaction_server ()
+	      && ((bufptr->flags & PGBUF_BCB_FLAGS_MASK) == PGBUF_BCB_FLUSH_NOT_NEEDED_FLAG)));
```

Comment:
Could you explain it briefly if you don't mind?

Reply:
@wrlawodms 

So, I might have gotten a few things wrong:
- I assumed that the flush daemon is **not** executed in PTS context - it is, see `log_daemons_init`
- `PGBUF_BCB_FLUSH_NOT_NEEDED_FLAG` was added specifically for PTS to inhibit flush daemon to flush pages because there is nowhere to flush those pages to
- so, in effect, the pages remain in 
- in effect, I think, the PTS flush daemon only works on temporary page buffer pages

Maybe another way of dealing with this is to actually make the PTS flush daemon perform this purging from the PTs's page buffer.
However, this raise a concern about race conditions:
  - (1) what if PTS replication must replicate a log-record which re-allocates the page
  - (2) but the PTS flush daemon has not flushed (ie: purged) the page yet from the page buffer

(1) If flush daemon purged the page from page buffer, and no other user transactions re-fetched the page from PS, the replication would skip applying the re-allocation log record.
(2) If flush daemon purged the page from page buffer, and if a subsequent user transaction re-fetched the page from PS, the page would go through the page synchronization mechanism and this situation would also be ok.
(3) If flush daemon did not manage to purge the page, and the replication would need to re-apply the log record re-allocating the page, this is the race condition and it will crash.

So the change above is a "cheap" solution for this. Not **the solution**, but a "cheap" one.

Reply:
Overall, I understand why you invalidate BCBs directly, not leveraging the flush daemon. That's what you're explaining, right? 

> I assumed that the flush daemon is not executed in PTS context - it is, see log_daemons_init

1. Do you mean `pgbuf_page_flush_daemon_init()` in `pgbuf_daemons_init()` by the flush daemon? Or do you really mean a log-related daemon in `log_daemons_init()`? It could be the checkpoint daemon.
I'm asking to clarify.

2. The reason to make the chage to the assert is that BCBs were not invalidated so, it didn't consider PGBUF_BCB_FLAGS_MASK turned on in PTS. Do I get it right?

Reply:
> 1. Do you mean `pgbuf_page_flush_daemon_init()` in `pgbuf_daemons_init()` by the flush daemon? Or do you really mean a log-related daemon in `log_daemons_init()`? It could be the checkpoint daemon.
> I'm asking to clarify.

You are correct. I am wrong.
Log daemons have nothing to do with this case. It is the page buffer daemons. And they are started in all server flavors (ATS, PTS, PS) regardless of any condition - `pgbuf_daemons_init`.

> 2. The reason to make the chage to the assert is that BCBs were not invalidated so, it didn't consider PGBUF_BCB_FLAGS_MASK turned on in PTS. Do I get it right? 

Yes, buffer control blocks were not invalidated before this change.
And, given your comment, I am starting to think that we will encounter further issues/crashes once we start testing without the entire database fitting into page buffer. I assume that - probably - >~97% of all tests up to now have only encountered the situation where the entire database fitted into memory and the page buffer flush daemons did not have much to do.
Or maybe I get this wrong and they also consider age and they did purge the page buffer (LRU? maybe). I don't know actually.

However, http://jira.cubrid.org/browse/LETS-596 is written to come back to this implementation once we are past a critical point that allows us to test further scenarios.

Reply:
> Or maybe I get this wrong and they also consider age and they did purge the page buffer (LRU? maybe). I don't know actually.

I did an analysis of the page buffer pool. From what I know, if there is no available BCB for a new page, a BCB that is basically not dirty is assigned for it (there are some other conditions such as not flushing or something). This is called "victimize". Which BCB is victimized is handled by basically multi-layered LRU lists. To keep enough victimizable BCBs, the page flush daemon, pgbuf_page_flush_daemon_task, periodically flushes BCBs. 
 If all BCB are not dirty, i.e. no victimizable BCB, a thread wanting to allocate a new BCB waits until a BCB gets victimizable by the flush daemon or others. A BCB that is assigned in this way is called a "direct victim".

Basically, I guess a data page is "purged" while victimizing. I hope this information helps you, @cristiarg.